### PR TITLE
add quick_sort, quick_select, insertion_sort

### PR DESCRIPTION
**Pull request by MoustaphaSaad**

### CODE REVIEW

File: cpprelude/inc/cpprelude/algorithm.h [hunk 1]
```diff
@@ -55,6 +96,59 @@ namespace cpprelude
...
+		auto it = begin_it;
+		for(usize i = 1; i < count; ++i)
+		{
+			it = next(it);
+
+			usize p_ix = i-1;
+			auto t_it = it;
+			auto t_p_it = prev(t_it);
+			
+			while(less_than(*t_it, *t_p_it))
```
Comment (NoraRo) [inline cpprelude/inc/cpprelude/algorithm.h]: will this insertion sort used only in small sizes of data? if not we can improve this algo. by using binary search.

File: cpprelude/inc/cpprelude/algorithm.h [hunk 2]
```diff
@@ -151,24 +245,281 @@ namespace cpprelude
...
+
+		return details::partition_result<iterator_type>(begin_it, 0);
+	}	
+
+	template<typename iterator_type, typename function_type = details::default_less_than<typename iterator_type::data_type>>
+	details::partition_3way_result<iterator_type>
+	_partition_3way(iterator_type begin_it, usize count, function_type less_than = function_type())
+	{
+		if (count == 0)
+			return details::partition_3way_result<iterator_type>(next(begin_it, count), count, next(begin_it), count);
```
Comment (NoraRo) [inline cpprelude/inc/cpprelude/algorithm.h]: Why calling next function? if count == 0 then next(begin_it,count) equals begin_it. right?

File: cpprelude/inc/cpprelude/algorithm.h [hunk 3]
```diff
@@ -151,24 +245,281 @@ namespace cpprelude
...
+		while(it_ix <= gt_ix)
 		{
+			bool it_l_k = less_than(*it, k_val);
+			bool k_l_it = less_than(k_val, *it);
 
-			if (less_than(*b ,*a))
-				return false;
+			if(it_l_k)
+			{
+				auto temp_value = tmp::move(*lt_it);
```
Comment (NoraRo) [inline cpprelude/inc/cpprelude/algorithm.h]: merge_sort needs to use tmp::move in its swapping values.

File: cpprelude/inc/cpprelude/algorithm.h [hunk 4]
```diff
@@ -151,24 +245,281 @@ namespace cpprelude
...
+		return _partition(begin_it, count, less_than).iterator;
+	}
+
+	template<typename iterator_type, typename function_type = details::default_less_than<typename iterator_type::data_type>>
+	void
+	_quick_sort(iterator_type begin_it, usize count, function_type less_than = function_type())
+	{
+		if (count <= 1)
+			return;
+
```
Comment (NoraRo) [inline cpprelude/inc/cpprelude/algorithm.h]: Use cutoff here too?

File: cpprelude/inc/cpprelude/algorithm.h (hunk 1 above)
Comment (MoustaphaSaad) [inline cpprelude/inc/cpprelude/algorithm.h]: yes we can improve it later to be used independently

File: cpprelude/inc/cpprelude/algorithm.h (hunk 2 above)
Comment (MoustaphaSaad) [inline cpprelude/inc/cpprelude/algorithm.h]: fixed

File: cpprelude/inc/cpprelude/algorithm.h (hunk 3 above)
Comment (MoustaphaSaad) [inline cpprelude/inc/cpprelude/algorithm.h]: task added to introduce swap function

File: cpprelude/inc/cpprelude/algorithm.h (hunk 4 above)
Comment (MoustaphaSaad) [inline cpprelude/inc/cpprelude/algorithm.h]: i think this is legacy and buggy so i think it should be removed

Comment (MoustaphaSaad) [inline cpprelude/inc/cpprelude/algorithm.h]: quick select has a bug since it depends on _partition ... duplicate keys make it stuck in infinite loop

**Pull request closed without merging**